### ISSUE TITLE
Incomplete variable declaration not well recovered

### ISSUE THREAD

**Issue by mickaelistria** [eclipse-jdtls/eclipse-jdt-core-incubator#1038]
Inspired from CompletionContextTests.test0136
```java
public class X {
  public void foo() {
    final ZZZZ
  }
}
```
is recovered "weakly" to DOM as
```java
package test;
public class X {
  public void foo(){
  }
}
```
without variable, while Javac has
```java
package test;

public class X {
    
    public void foo() {
        final ZZZZ <error>;
    }
}
```
It would be good to recover the Variable Declaration better.

**Comment by robstryker** [comment]
This appears to be a regression from https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/commit/6d1897301d6dba21cc37aedf39c28e58281fab51#diff-f6ffa5d531b4891f33f39aa1eff4b996c8b671ffb1fbdcddfc978a69f9bdbae4R2181

**Comment by mickaelistria** [comment]
It's indeed quite possible. The reason why some extra checks were added to remove some DOM node is because sometimes, Lombok or other can dynamically add non-existing node to the DOM and the best way I had found so far is checking some index condition. The conditions may not be good enough and may exclude some valid-ish nodes.
I hope the conditions can be refined to work for both excluding Lombok and including such erroneous node.
